insonmnia/gatekeeper/gatekeeper.go

```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("start gatekeeper instance",
+		zap.String("direction", cfg.Gatekeeper.Direction),
+		zap.Duration("delay", cfg.Gatekeeper.Delay),
+		zap.String("key", crypto.PubkeyToAddress(key.PublicKey).String()),
+		zap.String("day limit", keeper.DayLimit.String()),
+		zap.String("spent today", keeper.SpentToday.String()))
+
+	if keeper.DayLimit.Cmp(big.NewInt(0)) == 0 {
+		return nil, fmt.Errorf("used key is not keeper")
+	}
+
+	if keeper.Frozen {
+		return nil, fmt.Errorf("keeper with given key is frozen")
+	}
+
+	return &Gatekeeper{
+		key:    key,
+		cfg:    cfg,
+		in:     in,
+		out:    out,
+		logger: logger,
+		bch:    bch,
+	}, nil
+}
+
+func (g *Gatekeeper) Serve(ctx context.Context) error {
+	t := util.NewImmediateTicker(g.cfg.Gatekeeper.Period)
+
+	// straight load freezing time
+	g.loadFreezeTime(ctx)
+
+	freezingTimeLoader := util.NewImmediateTicker(g.cfg.Gatekeeper.ReloadFreezingPeriod)
+
+	for {
+		select {
+		case <-t.C:
+			g.processTransaction(ctx)
+		case <-freezingTimeLoader.C:
+			go g.loadFreezeTime(ctx)
+		}
+	}
+}
+
+func (g *Gatekeeper) processTransaction(ctx context.Context) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.logger.Debug("start transaction processing")
+
+	inTxs, outTxs, err := g.loadTransactions(ctx)
+	if err != nil {
+		return err
+	}
+
+	// find payin transactions doesn't exists in payout
+	for k, inTx := range inTxs {
+		_, ok := outTxs[k]
+		if !ok {
+			g.logger.Debug("found new unpaid transaction",
+				zap.String("from", inTx.From.String()),
+				zap.String("value", inTx.Value.String()),
+				zap.String("tx number", inTx.Number.String()),
+				zap.Uint64("block number", inTx.BlockNumber))
+
+			go g.processUnpaidTransaction(ctx, inTx)
+		}
+	}
+
+	g.findScummyTransactions(ctx, inTxs, outTxs)
+	g.logger.Debug("finish transaction processing")
+	return nil
+}
+
+func (g *Gatekeeper) loadTransactions(ctx context.Context) (map[string]*blockchain.GateTx, map[string]*blockchain.GateTx, error) {
+	var err error
+	inTxs, err := g.in.GetPayinTransactions(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	outTxs, err := g.out.GetPayoutTransactions(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	g.logger.Info("loaded transactions",
+		zap.Int("amount of payin transactions", len(inTxs)),
+		zap.Int("amount of payout transactions", len(outTxs)))
+
+	return inTxs, outTxs, nil
+}
+
+// checkDelay verify that tx out of delay
+func (g *Gatekeeper) checkDelay(ctx context.Context, tx *blockchain.GateTx) bool {
+	payinTimestamp, err := g.in.GetGateTransactionTime(ctx, tx)
+	if err != nil {
+		return false
+	}
+	// cast delay, time of payin tx and now time to uint64
+	payinTime := int64(payinTimestamp)
+	delay := int64(g.cfg.Gatekeeper.Delay.Seconds())
+	nowTime := time.Now().UTC().Unix()
+
+	g.logger.Debug("delay check", zap.Int64("delay", payinTime+delay), zap.Int64("nowTime", nowTime))
+	if nowTime < payinTime+delay {
```

Comment:
you can just return `payinTime+delay > nowTime`, but I'd prefer not calculate seconds by hand, instead use time.After, time.Before, time.Add methods.